packages/devtools_app/lib/src/debugger/program_explorer_model.dart

```diff
@@ -0,0 +1,512 @@
+        root.addChild(
+          VMServiceObjectNode(
+            controller,
+            field.name,
+            field,
+          ),
+        );
+      }
+    }
+
+    // Clear out the _childrenAsMap map.
+    root._trimChildrenAsMapEntries();
+
+    root._sortEntriesByType();
+    _outline = root.children;
+    return _outline;
+  }
+
+  /// Given a flat list of service protocol scripts, return a tree of scripts
+  /// representing the best hierarchical grouping.
+  static List<VMServiceObjectNode> createRootsFrom(
+    ProgramExplorerController controller,
+    List<VMServiceLibraryContents> libs,
+  ) {
+    // The name of this node is not exposed to users.
+    final root = VMServiceObjectNode(controller, '<root>', ObjRef(id: '0'));
+
+    for (var lib in libs) {
+      for (final script in lib.lib.scripts) {
+        _buildScriptNode(root, script, lib: lib.lib);
+      }
+    }
+
+    // Clear out the _childrenAsMap map.
+    root._trimChildrenAsMapEntries();
+
+    final processed = root.children
+        .map((e) => e._collapseSingleChildDirectoryNodes())
+        .toList();
+    root.children.clear();
+    root.addAllChildren(processed);
+
+    // Sort each subtree to use the following ordering:
+    //   - Scripts
+    //   - Classes
+    //   - Functions
+    //   - Variables
+    root._sortEntriesByType();
+
+    // Place the root library's parent node at the top of the explorer if it's
+    // part of a package. Otherwise, it's a file path and its directory should
+    // appear near the top of the list anyway.
+    final rootLib =
+        serviceManager.isolateManager.selectedIsolateState.isolateNow.rootLib;
+    if (rootLib.uri.startsWith('package:') ||
+        rootLib.uri.startsWith('google3:')) {
+      final parts = rootLib.uri.split('/')..removeLast();
+      final path = parts.join('/');
+      for (int i = 0; i < root.children.length; ++i) {
+        if (root.children[i].name.startsWith(path)) {
+          final rootLibNode = root.removeChildAtIndex(i);
+          root.addChildAtIndex(rootLibNode, 0);
+          break;
+        }
+      }
+    }
+
+    return root.children;
+  }
+
+  static VMServiceObjectNode _buildScriptNode(
+    VMServiceObjectNode node,
+    ScriptRef script, {
+    LibraryRef lib,
+  }) {
+    final parts = script.uri.split('/');
+    final name = parts.removeLast();
+
+    for (final part in parts) {
+      // Directory nodes shouldn't be selectable unless they're a library node.
+      node = node._getCreateChild(
+        part,
+        null,
+        isSelectable: false,
+      );
+    }
+
+    node = node._getCreateChild(name, script);
+    if (!node.isSelectable) {
+      node.isSelectable = true;
+    }
+    node.script = script;
+
+    // If this is a top-level node and a library is specified, this must be a
+    // library node.
+    if (parts.isEmpty && lib != null) {
+      node.object = lib;
+    }
+    return node;
+  }
+
+  VMServiceObjectNode _getCreateChild(
```

Comment:
nit: rename to lookupChild

Reply:
Will rename to `_lookupOrCreateChild` since it returns an existing child when possible, but falls back to creating a new one.